### PR TITLE
[#3112] Add supprt for TCP_INFO when using EpollSocketChannel

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

On linux you can gather various metrics using getsockopt(..., TCP_INFO...). We should support this in EpollSocketChannel.

Modifications:

Add EpollSocketChannel.tcpInfo() which return TcpInfo that holds all relevant metrics.

Result:

TCP_INFO is supported now.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@trustin ping :)

**Comment by trustin** [comment]
It depends on what users intend to do with tcpinfo. Why do you dislike reusing `EpollTcpInfo`?

**Comment by normanmaurer** [comment]
@trustin I think it is just a cleaner API to return a new object and not mutate an existing one. Also I think object creation is the "smaller" problem with calling it heavily as every call also needs to do a JNI Call, which is the "more expensive thing" IMHO.

If we really want to reuse TcpInfo we should also change the method name to `updateTcpInfo` or something like that. That said I'm still not convinced that it is a good thing to do.

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollTcpInfo.java [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+ * <p>
+ * struct tcp_info
+ * {
+ *      __u8    tcpi_state;
+ *      __u8    tcpi_ca_state;
+ *      __u8    tcpi_retransmits;
+ *      __u8    tcpi_probes;
+ *      __u8    tcpi_backoff;
+ *      __u8    tcpi_options;
+ *      __u8    tcpi_snd_wscale : 4, tcpi_rcv_wscale : 4;
```
Comment (normanmaurer) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollTcpInfo.java]: @trustin do we want to use byte for all u8 methods ? or better stick with int ?

Comment (trustin) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollTcpInfo.java]: Should be short or int.

**Comment by trustin** [comment]
Well, you could think the new method similarly to `toArray(T[])` in collections API?

**Comment by normanmaurer** [comment]
@trustin as you seems to feel very strong about it why not just keep both ?

**Comment by trustin** [comment]
I actually meant for keeping both.

**Comment by normanmaurer** [comment]
@trustin ok let me update the PR and ping you again for final review :)

**Comment by normanmaurer** [comment]
@trustin ok addressed... please one last review before I merge :)

File: transport-native-epoll/src/main/c/io_netty_channel_epoll_Native.h [hunk 2]
```diff
@@ -98,6 +98,7 @@ jint Java_io_netty_channel_epoll_Native_isBroadcast(JNIEnv *env, jclass clazz, j
 jint Java_io_netty_channel_epoll_Native_getTcpKeepIdle(JNIEnv *env, jclass clazz, jint fd);
 jint Java_io_netty_channel_epoll_Native_getTcpKeepIntvl(JNIEnv *env, jclass clazz, jint fd);
 jint Java_io_netty_channel_epoll_Native_getTcpKeepCnt(JNIEnv *env, jclass clazz, jint fd);
+void Java_io_netty_channel_epoll_Native_tcpInfo0(JNIEnv *env, jclass clazz, jint fd, jintArray array);
```
Comment (trustin) [inline transport-native-epoll/src/main/c/io_netty_channel_epoll_Native.h]: Consistency: `JNIEnv* env`

**Comment by trustin** [comment]
Looks great to me. Please cherry-pick after fixing asterisk position consistency.

**Comment by normanmaurer** [comment]
@trustin thanks... Cherry-picked into 4.0 (d79e4ffe071b7ac83692dfd4b08b7af6dfbc192e), 4.1 (c5bd8fd264986b1a3bdbaf5cb4488981ce554b78) and master (201d9ed9badaa70c941943268d974e61c4e84e9d)

**Pull request closed without merging**